src/plugins/data/public/search/search_service.ts

```diff
@@ -0,0 +1,112 @@
+import { ISearchSetup } from './i_search_setup';
+import { createAppMountSearchContext } from './create_app_mount_context_search';
+import { SYNC_SEARCH_STRATEGY, syncSearchStrategyProvider } from './sync_search_strategy';
+import {
+  TSearchStrategyProvider,
+  TRegisterSearchStrategyProvider,
+  TSearchStrategiesMap,
+} from './i_search_strategy';
+import { TStrategyTypes } from './strategy_types';
+import { esSearchService } from './es_search';
+
+/**
+ * Extends the AppMountContext so other plugins have access
+ * to search functionality in their applications.
+ */
+declare module 'kibana/public' {
+  interface AppMountContext {
+    search?: ISearchAppMountContext;
+  }
+}
+
+/**
+ * The search plugin exposes two registration methods for other plugins:
+ *  -  registerSearchStrategyProvider for plugins to add their own custom
+ * search strategies
+ *  -  registerSearchStrategyContext for plugins to expose information
+ * and/or functionality for other search strategies to use
+ *
+ * It also comes with two search strategy implementations - SYNC_SEARCH_STRATEGY and ES_SEARCH_STRATEGY.
+ */
+export class SearchService implements Plugin<ISearchSetup, void> {
+  /**
+   * A mapping of search strategies keyed by a unique identifier.  Plugins can use this unique identifier
+   * to override certain strategy implementations.
+   */
+  private searchStrategies: TSearchStrategiesMap = {};
+
+  /**
+   * Exposes context to the search strategies.
+   */
+  private contextContainer?: IContextContainer<TSearchStrategyProvider<any>>;
+
+  constructor(private initializerContext: PluginInitializerContext) {}
+
+  public setup(core: CoreSetup): ISearchSetup {
+    core.application.registerMountContext<'search'>('search', () => {
+      return createAppMountSearchContext(this.searchStrategies);
+    });
+
+    this.contextContainer = core.context.createContextContainer();
+
+    const registerSearchStrategyProvider: TRegisterSearchStrategyProvider = <
+      T extends TStrategyTypes
+    >(
+      plugin: PluginOpaqueId,
+      name: T,
+      strategyProvider: TSearchStrategyProvider<T>
+    ) => {
+      this.searchStrategies[name] = this.contextContainer!.createHandler(plugin, strategyProvider);
+    };
+
+    const api = {
+      registerSearchStrategyContext: this.contextContainer!.registerContext,
+      registerSearchStrategyProvider,
+    };
+
+    api.registerSearchStrategyContext(this.initializerContext.opaqueId, 'core', () => core);
+    api.registerSearchStrategyProvider(
+      this.initializerContext.opaqueId,
+      SYNC_SEARCH_STRATEGY,
+      syncSearchStrategyProvider
+    );
+
+    // ES search capabilities are written in a way that it could easily be a separate plugin,
+    // however these two plugins are tightly coupled due to the default search strategy using
+    // es search types.
+    esSearchService(this.initializerContext).setup(core, { search: api });
+
+    return api;
+  }
+
+  public start(core: CoreStart) {}
```

Comment:
hmm, how does this work ? setup() contract includes registerSearchStrategyContext and registerSearchStrategyProvider ... that makes sense

what about start method ? (i was expecting `search` function here)

Reply:
`search` functionality is  passed down via AppMountContext, or RouteHandlerContext. In NP land, I can't think of a situation they will need to be access via the `start` contract.  We exposed it in `Setup` just for legacy land.  We think we could expose in start too (or instead of). I don't think it really matters at that point.

Reply:
exposing `search` in setup method means that who ever uses it has no idea if searchStrategies have been registered yet.
not exposing `search` in start:  how can i pass search to any other plugin then ? my plugin would want to register esaggs function to expressions function registry, esaggs requires `search`, so we kind of need it in the contracts.

Reply:
also ... it doesn't seem we expose it in setup currently

Reply:
Yea, I honestly couldn't think of a reason to expose it, but you just gave one example.  It will be interesting though because esaggs will need to use it bound to the current user.

Mind if we punt on that for this PR and then we can work on adding it in for real to replace esaggs and make the adjustments needed that way?